neqo-transport/src/streams.rs

```diff
@@ -0,0 +1,464 @@
+            } => {
+                stats.max_streams += 1;
+                self.handle_max_streams(stream_type, maximum_streams);
+            }
+            Frame::DataBlocked { data_limit } => {
+                // Should never happen since we set data limit to max
+                qwarn!("Received DataBlocked with data limit {}", data_limit);
+                stats.data_blocked += 1;
+                self.handle_data_blocked();
+            }
+            Frame::StreamDataBlocked { stream_id, .. } => {
+                qtrace!("Received StreamDataBlocked");
+                stats.stream_data_blocked += 1;
+                // Terminate connection with STREAM_STATE_ERROR if send-only
+                // stream (-transport 19.13)
+                if stream_id.is_send_only(self.role) {
+                    return Err(Error::StreamStateError);
+                }
+
+                if let (_, Some(rs)) = self.obtain_stream(stream_id)? {
+                    rs.send_flowc_update();
+                }
+            }
+            Frame::StreamsBlocked { .. } => {
+                stats.streams_blocked += 1;
+                // We send an update evry time we retire a stream. There is no need to
+                // trigger flow updates here.
+            }
+            _ => unreachable!("This is not a stream RecoveryToken"),
+        }
+        Ok(())
+    }
+
+    pub fn write_frames(
+        &mut self,
+        priority: TransmissionPriority,
+        builder: &mut PacketBuilder,
+        tokens: &mut Vec<RecoveryToken>,
+        stats: &mut FrameStats,
+    ) -> Res<()> {
+        if priority == TransmissionPriority::Important {
+            // Send `DATA_BLOCKED` as necessary.
+            self.flow_control_sender
+                .borrow_mut()
+                .write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            // Send `MAX_DATA` as necessary.
+            self.flow_control_receiver
+                .borrow_mut()
+                .write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            self.recv_streams.write_frames(builder, tokens, stats)?;
+
+            self.remote_streams_fc[StreamType::BiDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+            self.remote_streams_fc[StreamType::UniDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            self.local_streams_fc[StreamType::BiDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
+
+            self.local_streams_fc[StreamType::UniDi].write_frames(builder, tokens, stats)?;
+            if builder.remaining() < 2 {
+                return Ok(());
+            }
```

Comment:
I would factor this out again into its own function.  Maybe `write_maintenance_frames`.